data/src/test/java/org/apache/iceberg/data/TestGenericTaskWriter.java

```diff
@@ -0,0 +1,151 @@
+
+import java.io.File;
+import java.io.IOException;
+import java.time.OffsetDateTime;
+import java.util.Collections;
+import java.util.List;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.AppendFiles;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.hadoop.HadoopCatalog;
+import org.apache.iceberg.io.OutputFileFactory;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.StructLikeSet;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestGenericTaskWriter {
+  @Rule public final TemporaryFolder temp = new TemporaryFolder();
+
+  private final Schema schema =
+      new Schema(
+          Types.NestedField.required(1, "level", Types.StringType.get()),
+          Types.NestedField.required(2, "event_time", Types.TimestampType.withZone()),
+          Types.NestedField.required(3, "message", Types.StringType.get()),
+          Types.NestedField.optional(
+              4, "call_stack", Types.ListType.ofRequired(5, Types.StringType.get())));
+
+  private final PartitionSpec spec =
+      PartitionSpec.builderFor(schema).hour("event_time").identity("level").build();
+
+  private Table table;
+  private List<Record> testRecords = Lists.newArrayList();
+  private final String testFileFormat;
+
+  @Parameterized.Parameters(name = "FileFormat = {0}")
+  public static Object[][] parameters() {
+    return new Object[][] {{"avro"}, {"orc"}, {"parquet"}};
+  }
+
+  public TestGenericTaskWriter(String fileFormat) throws IOException {
+    this.testFileFormat = fileFormat;
+  }
+
+  @Before
+  public void createTable() throws IOException {
+    File testWareHouse = temp.newFolder();
+    if (testWareHouse.exists()) {
+      Assert.assertTrue(testWareHouse.delete());
+    }
+
+    Catalog catalog = new HadoopCatalog(new Configuration(), testWareHouse.getPath());
+    this.table =
+        catalog.createTable(
+            TableIdentifier.of("logging", "logs"),
+            schema,
+            spec,
+            Collections.singletonMap("write.format.default", testFileFormat));
+  }
+
+  // test write and java API write data code demo
+  private void writeRecords() throws IOException {
+    GenericAppenderFactory appenderFactory =
+        new GenericAppenderFactory(table.schema(), table.spec());
+
+    FileFormat fileFormat =
+        FileFormat.valueOf(
+            table.properties().getOrDefault("write.format.default", "parquet").toUpperCase());
+
+    int partitionId = 1;
```

Comment:
@youngxinler What is the expected behavior here with partitionId being hard-coded?

I have tried to implement your code here in my own project, but it results in the partitions reporting as undefined instead of properly fanning out, at least when querying from Trino.

Reply:
* @param partitionId First part of the file name
   * @param taskId Second part of the file name
   * @param operationId Third part of the file name
used to generate file path.  Can you show me the code you used to write it?  `PartitionSpec` should be set table.spec().

Reply:
partitionId First part of the file name.
taskId Second part of the file name.
operationId Third part of the file name.
used to generate file path. Can you show me the code you used to write it? PartitionSpec should be set table.spec().

Reply:
I appreciate the info, turns out I was using the wrong constructor on this:
```
    GenericAppenderFactory appenderFactory =
        new GenericAppenderFactory(table.schema(), table.spec());
```
Had omitted the second argument. Works great now, thanks!